File: 424. Longest Repeating Character Replacement.cpp
class Solution {
public:
    int characterReplacement(string s, int k) {        
        if (s.length() == 0)
            return 0;
        vector<int> map(128, 0);        
        int begin = 0, end = 0, d = 0;
        int max_count = 0;
        while(end < s.length()){                              
            max_count = max(max_count, ++map[s[end]]);
            while (end - begin + 1 - max_count > k){
                map[s[begin]]--;
                begin++;
            }
            d = max(end - begin + 1, d);
            end++;
        }        
        return d;    
    }
};

File: Python/424. Longest Repeating Character Replacement.py
from collections import defaultdict
class Solution:
    def characterReplacement(self, s: str, k: int) -> int:
        start = 0
        counter = defaultdict(int)
        maxCount = 0
        maxLength = 0
        for i in range(len(s)):
            counter[s[i]] += 1
            maxCount = max(maxCount, counter[s[i]])
            if i - start + 1 - maxCount > k:
                counter[s[start]] -= 1
                start += 1            
            maxLength = max(i - start + 1, maxLength)
        return maxLength
